### ISSUE TITLE
Use non-Number values for data on StockLine chart type

### ISSUE THREAD

**Issue by BaroneX** [capitalone/react-native-pathjs-charts#4]
I want to show the xKey as a string,but it's a number, i tried to modified it,but it doesn't work.

**Comment by caledhwa** [comment]
I'll look at maybe changing this. Do you have some sample data you are trying to work with?

**Comment by cwxStrong** [comment]
why can't see the X and Y data
![image](https://cloud.githubusercontent.com/assets/17668835/18699173/2265a126-8002-11e6-87ca-8096d726be5b.png)

**Comment by marzolfb** [comment]
@BaroneX I'm confused about what you are trying to accomplish. 

`xkey` is a property you set on the `StockLine` component to identify the key name you use when specifying your data. So if you have this:

```
<StockLine data={sampleData.stockLine.data} options={sampleData.stockLine.options} xKey='x' yKey='y' />
```

you would want your data to look like this:

```
stockLine: {
        data: [
            [{
                "x": 0,
                "y": 47782
            }...
```

or, as another example, if you want to use more mathematical terms for the x & y coordinates and use "abscissa" for the x (horizontal) coordinate-part name and "ordinate" for the y (vertical) coordinate-part name, you would have code that looked like this:

```
<StockLine data={sampleData.stockLine.data} options={sampleData.stockLine.options} xKey='abscissa' yKey='ordinate' />
```

and your data would look like this:

```
stockLine: {
        data: [
            [{
                "abscissa": 0,
                "ordinate": 47782
            }...
```

but I think you are using the term `xkey` to refer to something you see visually (something you are trying to make appear on the x-axis?).

If this is still an outstanding issue, can you explain it differently? Do you have some sample code and data to illustrate what you are trying to do?

**Comment by marzolfb** [comment]
@cwxStrong That looks like a separate issue that you are having specifically with the `SmoothLine` component in the example app. I'm guessing maybe you have changed the options on the sample data and changed the `showLabels` property from true to false. If that doesn't address your issue, could you open a separate issue with some example code & data illustrating the issue?

**Comment by BaroneX** [comment]
I want show the month in x and number in y
```
stockLine: {
        data: [
            [{
                "x": "January",
                "y": 47782
            },{
                "x": "February",
                "y": 2323
            }...
```
and another data to show the time
```
stockLine: {
        data: [
            [{
                "x": "2016-10-10 15:00",
                "y": 47782
            },{
                "x": "2016-10-10 16:00",
                "y": 2323
            },{
                "x": "2016-10-10 17:00",
                "y": 2323
            }...
```

**Comment by marzolfb** [comment]
This project is based on [react-pathjs-chart](https://github.com/rsamec/react-pathjs-chart) and there is a built in assumption in that project that the x and y data values are specifically Number data types. For now, at least, we will not deviate from the basis project and keep things as they were designed. 

As for alternatives, you could continue to use the StockLine chart type and use numbers for the months of the year (1,2...12) or you could use a different chart type like Bar.

**Comment by BaroneX** [comment]
thkq.

**Comment by hoang-tranminh** [comment]
Hi, 
I have the same need, to show string labels on X axis instead of numbers. I found a work around, in which we still pass x values as numbers to the chart and we need to change source code of file Axis.js a little bit. 

In file Axis.js : 
` const ticks =_.map(axis.ticks, function (c, i) {
      const label = options.labelComponent !== undefined? React.cloneElement(options.labelComponent,{value:c}) : c
      let gxy = horizontal ? [scale(c),chartArea.y.min]:[chartArea.x.min,scale(c)]

      let returnValue = <G key={i} x={gxy[0]} y={gxy[1]}>

                {options.showTicks &&
                  <Circle r="2" cx="0" cy="0" stroke="grey" fill="grey" />
                }

                {options.showLabels &&
                  <Text x={xy[0]} y={xy[1]}
                        fontFamily={textStyle.fontFamily}
                        fontSize={textStyle.fontSize}
                        fontWeight={textStyle.fontWeight}
                        fontStyle={textStyle.fontStyle}
                        fill={textStyle.fill}
                        textAnchor={textAnchor}>
                        {label.toString()}
                  </Text>}
            </G>

      return returnValue
    })`

We can see that if we supply a "labelComponent" component, then this component will be used to render custom the axis label values, this component will take the current axis number value and return something (a string) to show as axis label.

I build a map, from number axis values to label strings, then supply the number array to the chart, but I can't figure out how to build a component that can show correctly with this  {label.toString()}, it always give me this sting "[object]".

So I updated 1 line of code in my local source to run:
` const ticks =_.map(axis.ticks, function (c, i) {
      **const label = options.labelFunction !== undefined? options.labelFunction(c) : c**
      let gxy = horizontal ? [scale(c),chartArea.y.min]:[chartArea.x.min,scale(c)]

      let returnValue = <G key={i} x={gxy[0]} y={gxy[1]}>`

Then can easily define a label function, that take in a axis number value and return a custom label string I want.

Please help me how to use a labelComponent proprerly.

**Comment by marzolfb** [comment]
I do want to look into this more but for now I just want to mention that this library is largely based on react-pathjs-charts and sometimes it is nice to look at some of the examples there for reference - https://github.com/rsamec/react-pathjs-chart/blob/95b56505e43c92eca1e7750b4070fac96358d7f0/example/src/ChartDemo.js#L48

**Comment by hoang-tranminh** [comment]
thanks

**Comment by eriksape** [comment]
in the options on xAxis is one value called tickValues

```js
axisX: {
                showAxis: true,
                showLines: false,
                showLabels: true,
                showTicks: false,
                zeroAxis: false,
                orient: 'bottom',
                tickValues: [
                  {value:'name1'},
                  {value:'name2'},
                  {value:'name3'},
                  {value:'name4'},
                  {value:'name5'},
                  {value:'name6'},
                  {value:'name7'}
                ],
```

it works but you need to modify Axis.js

on line 102
```js
let gxy = horizontal ? [scale(i),chartArea.y.min]:[chartArea.x.min,scale(i)]
```

**Comment by marzolfb** [comment]
@eriksape Your comment was super helpful and I thought was really close to a true solution. The problem is with your change to Axis.js, it "fixes" the StockLine chart but breaks the other chart types. A slight variation on your change to Axis.js that doesn't break for other chart types is this:

```
let scaleBase = isNaN(c) ? i : c
let gxy = horizontal ? [scale(scaleBase),chartArea.y.min]:[chartArea.x.min,scale(scaleBase)]
```

I will shoot for putting the changes to Axis.js in a new PR with a new example chart that shows non-numeric labels.

**Comment by minhquankq** [comment]
It will be better if we can pass a format function to format value. I have data by date time (datatype is long - timestamp), I want to display as date string.

**Comment by eriksape** [comment]
@minhquankq  for values you mean the yAxis?

it would be something like this

```js
axisY: {
                showAxis: true,
                showLines: false,
                showLabels: true,
                showTicks: false,
                zeroAxis: false,
                orient: 'bottom',
                tickValues: [
                  {value:'name1'},
                  {value:'name2'},
                  {value:'name3'},
                  {value:'name4'},
                  {value:'name5'},
                  {value:'name6'},
                  {value:'name7'}
                ],
```

**Comment by katscott** [comment]
To cover a wider range of use cases, things could be changed to handle an array of labels (i.e. tickValues with the axis fix) or a function to return a string for the label (formatted from something in the data).  There are certainly use cases where a single format function would be cleaner such as a date/time formatting with a time stamp or money format on a number value.

**Comment by marzolfb** [comment]
I think I've covered the desire for both static and dynamic axis tick labels as discussed in this thread in PR #48 - feel free to take a look at that and provide feedback if necessary.

**Comment by manjeets12** [comment]
error while updating property 'd' in shadow node of type :RNSVGPath 
with static tickValues, any idea what can be the reason
tickValues: [
              {value:'name1'},
              {value:'name2'},
              {value:'name3'},
              {value:'name4'},
              {value:'name5'},
              {value:'name6'},
              {value:'name7'},
              {value:'name8'},
              {value:'name9'},
              {value:'name10'},
              {value:'name11'},
              {value:'name12'},
              {value:'name13'},
              {value:'name14'},
              {value:'name15'},
              {value:'name16'},
              {value:'name17'},
              {value:'name18'},
              {value:'name19'},
              {value:'name20'},
              {value:'name21'}
        ]

**Comment by marzolfb** [comment]
That looks ok and similar to the example in the example app. Perhaps there is a problem elsewhere - what does the rest of your source look like?

**Comment by manjeets12** [comment]
Actually it was working fine without this, but as soon as I added this it breaks. `class LineGraph extends Component {
    constructor(props) {
        super(props);
    }

    render() {
    let data = [
      [{
        "x": -10,
        "y": -1000
      }, {
        "x": -9,
        "y": -729
      }, {
        "x": -8,
        "y": -512
      }, {
        "x": -7,
        "y": -343
      }, {
        "x": -6,
        "y": -216
      }, {
        "x": -5,
        "y": -125
      }, {
        "x": -4,
        "y": -64
      }, {
        "x": -3,
        "y": -27
      }, {
        "x": -2,
        "y": -8
      }, {
        "x": -1,
        "y": -1
      }, {
        "x": 0,
        "y": 0
      }, {
        "x": 1,
        "y": 1
      }, {
        "x": 2,
        "y": 8
      }, {
        "x": 3,
        "y": 27
      }, {
        "x": 4,
        "y": 64
      }, {
        "x": 5,
        "y": 125
      }, {
        "x": 6,
        "y": 216
      }, {
        "x": 7,
        "y": 343
      }, {
        "x": 8,
        "y": 512
      }, {
        "x": 9,
        "y": 729
      }, {
        "x": 10,
        "y": 1000
      }],
      [{
        "x": -10,
        "y": 100
      }, {
        "x": -9,
        "y": 81
      }, {
        "x": -8,
        "y": 64
      }, {
        "x": -7,
        "y": 49
      }, {
        "x": -6,
        "y": 36
      }, {
        "x": -5,
        "y": 25
      }, {
        "x": -4,
        "y": 16
      }, {
        "x": -3,
        "y": 9
      }, {
        "x": -2,
        "y": 4
      }, {
        "x": -1,
        "y": 1
      }, {
        "x": 0,
        "y": 0
      }, {
        "x": 1,
        "y": 1
      }, {
        "x": 2,
        "y": 4
      }, {
        "x": 3,
        "y": 9
      }, {
        "x": 4,
        "y": 16
      }, {
        "x": 5,
        "y": 25
      }, {
        "x": 6,
        "y": 36
      }, {
        "x": 7,
        "y": 49
      }, {
        "x": 8,
        "y": 64
      }, {
        "x": 9,
        "y": 81
      }, {
        "x": 10,
        "y": 100
      }]
    ]

    let options = {
      width: 280,
      height: 280,
      color: '#2980B9',
      margin: {
        top: 20,
        left: 45,
        bottom: 25,
        right: 20
      },
      animate: {
        type: 'delayed',
        duration: 200
      },
      axisX: {
        showAxis: true,
        showLines: true,
        showLabels: true,
        showTicks: true,
        zeroAxis: false,
        orient: 'bottom',
        label: {
          fontFamily: 'Arial',
          fontSize: 14,
          fontWeight: true,
          fill: '#34495E'
        },
        tickValues: [
              {value:'name1'},
              {value:'name2'},
              {value:'name3'},
              {value:'name4'},
              {value:'name5'},
              {value:'name6'},
              {value:'name7'},
              {value:'name8'},
              {value:'name9'},
              {value:'name10'},
              {value:'name11'},
              {value:'name12'},
              {value:'name13'},
              {value:'name14'},
              {value:'name15'},
              {value:'name16'},
              {value:'name17'},
              {value:'name18'},
              {value:'name19'},
              {value:'name20'},
              {value:'name21'}
        ],
      },
      axisY: {
        showAxis: true,
        showLines: true,
        showLabels: true,
        showTicks: true,
        zeroAxis: false,
        orient: 'left',
        label: {
          fontFamily: 'Arial',
          fontSize: 14,
          fontWeight: true,
          fill: '#34495E'
        }
      }
    }
        return (
            <View style ={{width:width,height:height}}>
                <SmoothLine data={data} options={options} xKey='x' yKey='y' />
            </View>
        )
    }


}`

**Comment by marzolfb** [comment]
@manjeets12 It turns out this behavior is the same for your example and the `StockLineChartStaticTickLabels` chart in the example app. I also discovered the cause of this behavior is the problem @qiangspecial pointed out in #89. It's fixed in #91
